lib/rules/no-nested-landmark.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const Rule = require('./base');
+const AstNodeInfo = require('../helpers/ast-node-info');
+
+function createErrorMessage(element) {
+  return `Nested landmark elements on <${element}> detected. Landmark elements should not be nested within landmark elements of the same name.`;
+}
+
+const LANDMARK_ELEMENTS = ['header', 'main', 'aside', 'form', 'main', 'nav', 'footer'];
+const ROLES = ['banner', 'main', 'complementary', 'form', 'search', 'navigation', 'contentinfo'];
+
+module.exports = class NoNestedLandmark extends Rule {
+  visitor() {
+    this._parentLandmarkNode = null;
```

Comment:
I think we can remove the manual enter/exit handling in favor of the (newish) support for paths passed to the hooks.

Pseudo code:

```js
return {
  ElementNode(node, path) {
    for (let parent of path.parents()) {
      if (this.isLandmarkElement(parent.node)) {
        this.log(...);
        return;
      }
    }
  },
};
```

For example:

https://github.com/ember-template-lint/ember-template-lint/blob/26fcdea982bb6f1b2050679f9f80ff211c238f33/lib/rules/require-input-label.js#L8-L19

Reply:
The way path.parents() works is that it only capture the immediate parents of the current node not the ancestors I guess. I tried with the following logic:
```js
return {
  ElementNode(node, path) {
    for (let parent of path.parents()) {
      if (this.isLandmarkElement(parent.node)) {
        this.log(...);
        return;
      }
    }
  },
};


```
The above logic works for:
```html
<main>
  <main></main>
</main>
```
But not for:
```html
<main>
  <div>
    <main></main>
  </div>
</main>
```

Reply:
> Actually I took the cues from `no-nested-interactive`

Ya, but that was written before we can walk the parents 😄, and should be rewritten / refactored to avoid the error prone manual tracking.

Reply:
> The way path.parents() works is that it only capture the immediate parent of the current node, not the ancestors I guess.

Nope, it traverses all parents. Using the snippet you mentioned (main > div > main) and using `path.parents()` from the nested `main` _can **definitely**_ see the top level parent `main`. See this AST explorer demo:

https://astexplorer.net/#/gist/9e4b550991b7b62eaa62f847dd09bf93/f050ca2a7b46a78dbae5d650ace1e5ac8fd628a1

Reply:
Thanks @rwjblue that worked, my bad I missed the condition to check whether the current child node is a landmark element or not, that's why my tests failed. Now after changing to the ElementNode hook, the code looks much cleaner and concise. 👍